### DISCUSSION TITLE
How to unit test entity classes?

### DISCUSSION THREAD

**Discussion by bolqqq** [mikro-orm/mikro-orm#4948] [General]
So we have a guideline in our company to cover everything with unit tests 100%.
Our Tech Leads decided to use MikroORM for future products any I have the task to build an example project with various use cases and how to test them.
I'm using the following example entitiy 
```
import { Entity, PrimaryKey, Property } from '@mikro-orm/core';

@Entity()
export class Author {
  @PrimaryKey()
  id!: number;

  @Property()
  createdAt: Date = new Date();

  @Property({ onUpdate: () => new Date() })
  updatedAt: Date = new Date();

  @Property()
  name!: string;

  @Property()
  email!: string;

  @Property({ nullable: true })
  age?: number;

  @Property()
  termsAccepted: boolean = false;

  @Property({ nullable: true })
  identities?: string[];

  @Property({ nullable: true })
  born?: Date;

  @Property({ version: true })
  version!: number;

  constructor(name: string, email: string) {
    this.name = name;
    this.email = email;
  }
}
```

I now need to write a unit test for this that covers this file 100%...
```
let orm: MikroORM;
let em: EntityManager;

describe('Author Entity', () => {
  beforeAll(async () => {
    orm = await MikroORM.init({ ...options, connect: false });
    em = orm.em.fork();
  });

  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
    em.clear();
  });

  test('should create an instance of Author', () => {
    // Arrange
    const testObj = { name: 'Tony Stark', email: 'tony@stark.industries' };

    // Act
    const author = new Author(testObj.name, testObj.email);

    // Assert
    expect(author).toBeInstanceOf(Author);
    expect(author.name).toBe(testObj.name);
    expect(author.email).toBe(testObj.email);
  });

  test('should update updatedAt of Author if object is updated', async () => {
    // Arrange
    const testObj = { name: 'Tony Stark', updatedName: 'Anthony Edward Stark', email: 'tony@stark.industries' };
    const createdTimestamp = '2023-01-01T00:00:00.000Z';
    const updatedTimestamp = '2023-01-02T00:00:00.000Z';

    // Act
    // Create initial author
    jest.setSystemTime(new Date(createdTimestamp));
    const author = new Author(testObj.name, testObj.email);
    await em.persistAndFlush(author);

    // Update author
    jest.setSystemTime(new Date(updatedTimestamp));
    author.name = testObj.updatedName;
    await em.persistAndFlush(author);

    // reload author
    const updatedAuthor = await em.findOne(Author, { id: author.id });

    // Assert
    expect(updatedAuthor).toBeInstanceOf(Author);
    expect(updatedAuthor!.name).toBe(testObj.updatedName);
    expect(updatedAuthor!.email).toBe(testObj.email);
    expect(updatedAuthor!.updatedAt).toStrictEqual(new Date(updatedTimestamp));
  });
});
```
This first test covers everything besides the following line in the entity `@Property({ onUpdate: () => new Date() })`

The second test doesn't work as I have no access to the database for persistAndFlush but I'm unsure how to test this at all. I haven't found anything online or in the documentation of MikroORM for this so maybe someone here can help me? I'm clueless how to mock or trigger. the onupdate behavior (or maybe even relations in the future) in a unit test.

**Comment by B4nan** [comment]
> `@Property({ onUpdate: () => new Date() })`

First of all, I would suggest you confirm this with your tech leads, as asking for 100% unit test coverage of entity file seems just weird. I can understand 100% branch coverage for your whole test suite (I try to pursue that in this repository myself), but it does not make any sense to have a unit test which would result in 100% coverage of some file, those things should be tested via integration tests. Also 100% line coverage is not the same as 100% branch coverage, this line will surely be interpreted = covered, what won't be covered is the `() => new Date()` callback - and that is about branch coverage, which they might not care that much.

> The second test doesn't work as I have no access to the database for persistAndFlush but I'm unsure how to test this at all.

If you really need/want to unit test this without using a database (I'd strongly discourage that, it does very little sense to me, and will add a lot of tight-coupling to your tests, they could also potentially break with minor/patch ORM releases, and it will be a pain in the ass to migrate major versions most probably), you will need to mock the internals of the ORM. And at this moment, I'd ask myself what I actually want to test here, as you would be just replicating the things that the ORM is testing already. For one missing branch, you will need to do _a lot_ (not just a lot of code, but also a lot of efforts into understanding the ORM internals). Theoretically, mocking the `Connection.execute()` method is one way, preferably via `jest.spyOn`, that could do it for this simple use case, but you will need to mimic how the SQL execution works for different query types (insert/update/select).

Instead, a much better way is to use an actual database, if your setup allows it, many people use in-memory sqlite for such tests. Nowadays, sqlite is pretty powerful and your entities should mostly work with that unless you use some dialect-specific mapping or raw defaults.

Or just use your actual driver with a real database, and cover those things in integration tests instead.

>     // reload author
>    const updatedAuthor = await em.findOne(Author, { id: author.id });

FYI this line is not reloading anything, `findOne` by a PK will just give you back the entity from the identity map (if present), without touching the database (it won't fire any query here). You'd have to use `em.refresh(author)` or add `refresh: true` to the options parameter.
